### PR TITLE
Restoring lost information to Drag and Drop doc

### PR DESCRIPTION
**Pull request by mhoeber**
DOC-1101

### CODE REVIEW

**Comment by mhoeber** [comment]
@srpearce  can you do quick review.  I think the the whole drag and drop doc can be redone, it's hard to parse for new people.  But scope of this is to just restore some info we inadvertently lost.

Pointing Melanie to this, can't seem to find name in github

**Comment by mhoeber** [comment]
@JAAkana

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 1]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+      'targets':  ['t5_c', 't6_c'],
+      'rule': 'anyof'
+    },
+    {
+      'draggables': ['1', '2'],
+      'targets': ['t2_h', 't3_h', 't4_h', 't7_h', 't8_h', 't10_h'],
+      'rule': 'anyof'
+    }]
+
+Draggables is the list of draggable IDs. Target is thelist of target IDs that
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: Add quotation marks around "Draggables" and "Target"

"thelist" -> "the list"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 2]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+        <target {attribute list} />
+        <target {attribute list} />
+        <target {attribute list} />
+        ...
+    </draggable>
+
+The attribute list in the tags above (``draggable`` and ``target``) is the same
+as for normal ``draggable`` and ``target`` tags. The only difference is when
+you will be specifying inner target position coordinates. Use the ``x`` and
+``y`` attributes you to set the offset of the inner target from the upper-left
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "attributes to set" (delete "you")

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 3]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+=====================================
+Limitations of targets on draggables
+=====================================
+
+* Currently there is a limitation to the level of nesting of targets.
+
+  Even though you can pile up a large number of draggables on targets that
+  themselves are on draggables, the Drag and Drop problem will be graded only
+  in the case if there is a maximum of two levels of targets. The first level
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: Strike "in the case" (or change "if" to "when")

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 4]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+  ensure that targets on draggables are smaller than the actual parent
+  draggable. Technically this is not necessary, but from the usability
+  perspective it is desirable.
+
+* You can have targets on draggables only in the case when there are base
+  targets defined (base targets are attached to the base image).
+
+  If you do not have base targets, then you can only have a single level of
+  nesting (draggables on the base image). In this case the client side will be
+  reporting (x,y) positions of each draggables on the base image.
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: draggables -> draggable

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 5]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+**********************
+Correct answer format
+**********************
+
+For specifying answers for targets on draggables, see `Answer format for targets on draggables`_.
+
+There are two correct answer formats: short and long
+
+If short the correct answer is mapping of ``draggable_id`` to ``target_id``::
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "If short the correct" -> "In short form, the correct"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 6]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+      'rule': 'anyof'
+    },
+    {
+      'draggables': ['1', '2'],
+      'targets': ['t2_h', 't3_h', 't4_h', 't7_h', 't8_h', 't10_h'],
+      'rule': 'anyof'
+    }]
+
+Draggables is the list of draggable IDs. Target is thelist of target IDs that
+draggables must be dragged to with considering rule. Rule is string.
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: I'm not sure what "must be dragged to with considering rule" means - ?

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 7]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+    correct_answer = [
+    {
+      'draggables':   ['7', '8'],
+      'targets':  ['tartget1', 'target2'],
+      'rule': 'unordered_equal'
+    }]
+
+
+* ``anyof``: Allows draggables to be dragged to any of targets. For students to
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "any of targets" -> "any target" or "any of the targets"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 8]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+  drag 7 and 8 to target1 or target2, which means that if 7 is on target1 and 8
+  is on target1 or 7 on target2 and 8 on target2 or 7 on target1 and 8 on
+  target2. Any of theese are correct with the `anyof` rule::
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: The second sentence in this paragraph ("For students to drag 7 and 8...") is making my head spin. Can we clarify it a little?

Also, "theese" -> "these"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 9]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+    {
+      'draggables':   ['7', '8'],
+      'targets':  ['tartget1', 'target2'],
+      'rule': 'anyof'
+    }]
+
+If ``can_reuse`` is true, then you have draggables a,b,c and 10 targets. These
+will allow you to drag 4 ``a`` draggables to [``target1``,  ``target4``,
+``target7``, ``target10``]; you do not need to write ``a`` four times. Also
+this will allow you to drag ``b`` draggable to target2 or target5 for target5
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "to drag _the_ `b` draggable"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 10]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+          'targets': ['target2', 'target5', 'target8'],
+          'rule': 'anyof'
+        },
+        {
+          'draggables': ['c'],
+          'targets': ['target3', 'target6', 'target9'],
+          'rule': 'unordered_equal'
+        }]
+
+Sometimes you want to allow students to drag only two ``b`` draggables, in this
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "draggables, in this" -> "draggables. In this"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 11]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+case you should use ``anyof+number`` of ``unordered_equal+number`` rule::
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "of" -> "in the" (I think?)

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 12]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+          'rule': 'anyof+number'
+        },
+        {
+          'draggables': ['c'],
+          'targets': ['target3', 'target6', 'target9'],
+          'rule': 'unordered_equal'
+        }]
+
+When there are no multiple draggables per targets (one_per_target=``true``),
+for same number of draggables, ``anyof`` is equal to ``unordered_equal``.
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "for the same"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 13]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+If ``can_reuse=true``, then one must use only the long form of correct answer.
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "of the correct answer"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 14]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+Answer format for targets on draggables
+=======================================
+
+As with the cases described above, an answer must provide precise positioning
+for each draggable (on which targets it must reside). In the case when a
+draggable must be placed on a target that itself is on a draggable, then the
+answer must contain the chain of target-draggable-target. 
+
+For example, suppose we have three draggables - ``up``, ``s``, and ``p``.
+Draggables ``s``, and ``p`` have targets on themselves. More specifically,
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: Strike comma after `s`

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 15]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+``p`` has three targets - ``1``, ``2``, and ``3``. The first requirement is
+that ``s``, and ``p`` are positioned on specific targets on the base image. The
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: Strike comma after `s`

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 16]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+    ]
+
+Note that you must specify rules for all draggables, even if a draggable gets
+included in more than one chain.
+
+*************
+Grading logic
+*************
+
+#. The students answer and the correct answer are parsed to the same format::
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "student's"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 17]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+    group_id: group_draggables, group_targets, group_rule
+
+  ``group_id`` is ordinal number, for every dict in correct answer incremental
+  ``group_id`` is assigned: 0, 1, 2, ...
+
+  Draggables from user answer are added to same group_id where identical
+  draggables from correct answer are, for example::
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "the user answer"
"the same group_id"
"the correct answer"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 18]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+    If correct_draggables[group_0] = [t1, t2] then
+    user_draggables[group_0] are all draggables t1 and t2 from user answer:
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "the user answer"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 19]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+2. For every group from user answer, for that group draggables, if ``number``
+   is in  group rule, set() is applied, if ``number`` is not in rule, set is
+   not applied::
+
+    set() : [t1, t2, t3, t3] -> [t1, t2, ,t3]
+
+  For every group, at this step, draggables lists are equal.
+
+3. For every group, lists of targets are compared using rule for that group.
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "using the rule"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 20]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+  ``group_id`` is assigned: 0, 1, 2, ...
+
+  Draggables from user answer are added to same group_id where identical
+  draggables from correct answer are, for example::
+
+    If correct_draggables[group_0] = [t1, t2] then
+    user_draggables[group_0] are all draggables t1 and t2 from user answer:
+    [t1] or [t1, t2] or [t1, t2, t2] etc..
+
+2. For every group from user answer, for that group draggables, if ``number``
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "the user answer"
"group's draggables"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 21]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+   is in  group rule, set() is applied, if ``number`` is not in rule, set is
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "is in the group rule, set() is applied. If `number` is not in the rule, set() is"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 22]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+  For every group, at this step, draggables lists are equal.
+
+3. For every group, lists of targets are compared using rule for that group.
+
+==========================
+Set and ``+number`` cases
+==========================
+
+``set()`` and ``+number`` are needed only for case of reusable draggables,
+for other cases there are no equal draggables in list, so set() does nothing.
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "... only for _the_ case of reusable draggables. For other cases there are no equal draggables in the list"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 23]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+* The ``set()`` operation allows you to create rule for the case of "any
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "a rule"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 24]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+  number of same draggable can be dragged to some targets"::
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "number of the same draggable can be dragged to multiple targets"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 25]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+    {
+      'draggables': ['draggable_1'],
+      'targets': ['target3', 'target6', 'target9'],
+      'rule': 'anyof'
+    }
+
+* The ``number`` rule is used for the case of reusable draggables, when one
+  want to fix number of draggable to drag. In this example only two instances
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst:594]: "wants to fix the number of draggables"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 26]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+    {
+      'draggables': ['draggable_1', 'draggable_1'],
+      'targets': ['target3', 'target6', 'target9'],
+      'rule': 'anyof+number'
+    }
+
+
+* Note, that in using rule ``exact``, one does not need ``number``, because you
+  cannot recognize from user interface which reusable draggable is on which
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "from the user interface"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 27]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+  target. For example::
+
+    {
+      'draggables': ['draggable_1', 'draggable_1', 'draggable_2'],
+      'targets': ['target3', 'target6', 'target9'],
+      'rule': 'exact'
+    }
+
+
+    Correct handling of this example is to create different rules for
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst:615]: Add a period

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 28]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+    draggable_1 and draggable_2
+
+* For ``unordered_equal`` (or ``exact``) we don't need ``number`` if you have
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "we" -> "you"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 29]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+  only the same draggable in the group, as the target length will provide
+  constraint for the number of draggables::
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: the constraint

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 30]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+
+    {
+      'draggables': ['draggable_1'],
+      'targets': ['target3', 'target6', 'target9'],
+      'rule': 'unordered_equal'
+    }
+
+  This means that only ``draggable_1`` can be dragged.
+
+* But if you have more that one different reusable draggable in list, you may
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "that" -> "than" 

"in the list"

File: en_us/shared/exercises_tools/drag_and_drop.rst [hunk 31]
```diff
@@ -250,7 +265,330 @@ If you don't specify a target, a student can drop a draggable item anywhere on t
...
+  use ``number`` rule::
```
Comment (srpearce) [inline en_us/shared/exercises_tools/drag_and_drop.rst]: "use the `number` rule"

**Comment by srpearce** [comment]
OK, quick-and-dirty proofing pass done. Wow.
